packages/multichain/src/adapters/caip-permission-adapter-eth-accounts.ts

```diff
@@ -0,0 +1,105 @@
+import {
+  type CaipAccountId,
+  type Hex,
+  KnownCaipNamespace,
+  parseCaipAccountId,
+} from '@metamask/utils';
+
+import type { Caip25CaveatValue } from '../caip25Permission';
+import { mergeScopes } from '../scope/transform';
+import type { ScopesObject, ScopeString } from '../scope/types';
+import { parseScopeString } from '../scope/types';
+
+const isEip155ScopeString = (scopeString: ScopeString) => {
+  const { namespace, reference } = parseScopeString(scopeString);
+
+  return (
+    namespace === KnownCaipNamespace.Eip155 ||
+    (namespace === KnownCaipNamespace.Wallet &&
+      reference === KnownCaipNamespace.Eip155)
+  );
+};
+
+export const getEthAccounts = (caip25CaveatValue: Caip25CaveatValue) => {
+  const ethAccounts: string[] = [];
+  const sessionScopes = mergeScopes(
+    caip25CaveatValue.requiredScopes,
+    caip25CaveatValue.optionalScopes,
+  );
+
+  Object.entries(sessionScopes).forEach(([_, { accounts }]) => {
+    accounts?.forEach((account) => {
+      const { address, chainId } = parseCaipAccountId(account);
+
+      if (isEip155ScopeString(chainId)) {
+        ethAccounts.push(address);
+      }
+    });
+  });
+
+  return Array.from(new Set(ethAccounts));
+};
+
+const setEthAccountsForScopesObject = (
+  scopesObject: ScopesObject,
+  accounts: Hex[],
+) => {
+  const updatedScopesObject: ScopesObject = {};
+
+  Object.entries(scopesObject).forEach(([scopeString, scopeObject]) => {
+    const { namespace, reference } = parseScopeString(scopeString);
+
+    const isWalletNamespace =
+      namespace === KnownCaipNamespace.Wallet && reference === undefined;
+
+    if (
+      !isEip155ScopeString(scopeString as ScopeString) &&
+      !isWalletNamespace
+    ) {
+      updatedScopesObject[scopeString as ScopeString] = scopeObject;
+      return;
+    }
+
+    const caipAccounts = accounts.map(
+      (account) =>
+        (isWalletNamespace
+          ? `wallet:eip155:${account}`
+          : `${scopeString}:${account}`) as CaipAccountId,
+    );
+
+    updatedScopesObject[scopeString as ScopeString] = {
+      ...scopeObject,
+      accounts: caipAccounts,
+    };
+  });
+
+  return updatedScopesObject;
+};
+
+// This helper must be called with existing eip155 scopes
+export const setEthAccounts = (
+  caip25CaveatValue: Caip25CaveatValue,
+  accounts: Hex[],
+) => {
+  return {
+    ...caip25CaveatValue,
+    requiredScopes: setEthAccountsForScopesObject(
+      caip25CaveatValue.requiredScopes,
+      accounts,
+    ),
+    optionalScopes: setEthAccountsForScopesObject(
+      {
+        wallet: {
+          methods: [],
+          notifications: [],
+        },
+        'wallet:eip155': {
```

Comment:
Total nit: Should these be `const`s somewhere

Reply:
https://github.com/MetaMask/core/pull/4784/commits/941849da0e93d4be7da0e771ffa464a957dba306